tests/Unit/Jobs/UpdateAllProductsTest.php

```diff
@@ -64,68 +61,216 @@ public function setUp(): void {
 		$this->assertEquals( self::JOB_NAME, $this->job->get_name() );
 	}
 
-	public function test_schedule_schedules_a_single_batched_job_with_items() {
-		$items = array(
-			$this->generate_simple_product_mock(),
-			$this->generate_simple_product_mock(),
-			$this->generate_simple_product_mock(),
-			$this->generate_simple_product_mock(),
-		);
-
-		$item_ids = array_map(
-			function ( $item ) {
-				return $item->get_id();
-			},
-			$items
-		);
+	/**
+	 * Test single empty batch to make sure it does not continue scheduling and calls handle_complete
+	 * by asserting `has_scheduled_action` was called only once when scheduling the batch and did not continue.
+	 */
+	public function test_single_empty_batch() {
+		$filtered_product_list = new FilteredProductList( [], 0 );
+		$this->action_scheduler->expects( $this->once() )
+			->method( 'schedule_immediate' )
+			->with( self::CREATE_BATCH_HOOK, [ 1 ] );
+		$this->product_repository->expects( $this->once() )
+			->method( 'find_sync_ready_products' )
+			->willReturn( $filtered_product_list );
 
-		$this->filtered_product_list->expects( $this->never() )
-			->method( 'get' );
+		/*
+		 * We expect only single call to `has_scheduled_action` when scheduling
+		 * the batch and no calls further since batch is empty.
+		 */
+		$this->action_scheduler->expects( $this->once() )
+			->method( 'has_scheduled_action' );
 
-		$this->filtered_product_list->expects( $this->once() )
-			->method( 'get_product_ids' )
-			->willReturn( $item_ids );
+		$this->job->schedule();
 
-		$this->filtered_product_list->expects( $this->once() )
-			->method( 'count' )
-			->willReturn( count( $items ) );
+		do_action( self::CREATE_BATCH_HOOK, 1 );
+	}
 
-		$this->filtered_product_list->expects( $this->exactly( 2 ) )
-			->method( 'get_unfiltered_count' )
-			->willReturn( count( $items ) );
+	/**
+	 * We test single half filled batch to make sure it stops after processing all the items
+	 */
+	public function test_single_partial_batch() {
+		$filtered_product_list = new FilteredProductList( $this->generate_simple_product_mocks_set( 1 ), 1 );
 
 		$this->action_scheduler->expects( $this->exactly( 2 ) )
-			->method( 'has_scheduled_action' )
+			->method( 'schedule_immediate' )
 			->withConsecutive(
-				array( self::CREATE_BATCH_HOOK, [ 1 ] ),
-				array( self::PROCESS_ITEM_HOOK, [ $item_ids ] )
-			)
-			->willReturnOnConsecutiveCalls( false, false );
+				[ self::CREATE_BATCH_HOOK, [ 1 ] ],
+				[ self::PROCESS_ITEM_HOOK, [ $filtered_product_list->get_product_ids() ] ]
+			);
+
+		$this->product_repository->expects( $this->once() )
+			->method( 'find_sync_ready_products' )
+			->willReturn( $filtered_product_list );
+
 		$this->action_scheduler->expects( $this->exactly( 2 ) )
+			->method( 'has_scheduled_action' );
+
+		$this->job->schedule();
+
+		do_action( self::CREATE_BATCH_HOOK, 1 );
+	}
+
+	/**
+	 * We test two filly loaded batches to make sure batch and process item jobs are correctly scheduled
```

Comment:
Just a small typo `fully loaded`. Also when we refer to batch and process items, these are separate actions within the full job.
```suggestion
	 * We test two fully loaded batches to make sure batch and process item actions are correctly scheduled
```

Reply:
Fixed the typo and changed jobs to actions. Thanks!